Remove -bounds-checking-unique-traps (replace with -fno-sanitize-merge=local-bounds)

thurstond opened the pull request:
#120613 removed -ubsan-unique-traps and replaced it with  -fno-sanitize-merge (introduced in #120511), which allows fine-grained control of which UBSan checks to prevent merging. This analogous patch removes -bound-checking-unique-traps, and allows it to be controlled via -fno-sanitize-merge=local-bounds.

Most of this patch is simply plumbing through the compiler flags into the bounds checking pass.

Note: this patch subtly changes -fsanitize-merge (the default) to also include -fsanitize-merge=local-bounds. This is different from the previous behavior, where -fsanitize-merge (or the old -ubsan-unique-traps) did not affect local-bounds (requiring the separate -bounds-checking-unique-traps). However, we argue that the new behavior is more intuitive.

Removing -bounds-checking-unique-traps and merging its functionality into -fsanitize-merge breaks backwards compatibility; we hope that this is acceptable since '-mllvm -bounds-checking-unique-traps' was an experimental flag.

thurstond committed:
Remove -bounds-checking-unique-traps (replace with -fno-sanitize-merge=local-bounds)

-fno-sanitize-merge (introduced in #120511) combines the functionality of -ubsan-unique-traps
and -bounds-checking-unique-traps, while allowing fine-grained control
of which UBSan checks to prevent merging. #120613 removed -ubsan-unique-traps.
This patch removes -bound-checking-unique-traps, which can be controlled
via -fno-sanitize-merge=local-bounds.

Note: this patch subtly changes -fsanitize-merge (the default) to also
include -fsanitize-merge=local-bounds. This is different from the
previous behavior, where -fsanitize-merge (or the old
-ubsan-unique-traps) did not affect local-bounds (requiring the separate
-bounds-checking-unique-traps). However, we argue that the new behavior
is more intuitive.

Removing -bounds-checking-unique-traps and merging its functionality into -fsanitize-merge breaks backwards compatibility; we hope that this is acceptable since '-mllvm -bounds-checking-unique-traps' was an experimental flag.

thurstond committed:
clang-format

vitalybuka commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -488,8 +489,11 @@ Removed Compiler Flags
   it's still in use.
-- The experimental flag '-ubsan-unique-traps' has been removed. It is
-  superseded by '-fno-sanitize-merge'.
+- The experimental flags '-ubsan-unique-traps' and
```
this is -mllvm flag, they don't worth mentioning

thurstond replied:
Removed

vitalybuka commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -445,9 +445,10 @@ New Compiler Flags
-- '-fsanitize-merge' (default) and '-fno-sanitize-merge' have been added for
-  fine-grained control of which UBSan checks are allowed to be merged by the
-  backend (for example, -fno-sanitize-merge=bool,enum).
+- ``-fsanitize-merge`` (default) and ``-fno-sanitize-merge`` have been added for
```
there is a special section "Sanitizers" on line 1154, maybe better to keep there?

thurstond replied:
Moved

vitalybuka commented on llvm/test/Instrumentation/BoundsChecking/runtimes.ll:
```diff
@@ -5,6 +5,21 @@
 ; RUN: opt < %s -passes='bounds-checking<min-rt-abort>' -S | FileCheck %s --check-prefixes=MINRTABORT
+;
+; merge defaults to true
+; RUN: opt < %s -passes='bounds-checking<merge=true>'              -S | FileCheck %s --check-prefixes=TR
```
I think one with trap and one with runtime is enough
we will need later one flag for guard, we don't want 3x them

thurstond replied:
Ok, trimmed the tests

vitalybuka commented on llvm/lib/Passes/PassBuilder.cpp:
```diff
@@ -1281,31 +1281,43 @@ parseRegAllocFastPassOptions(PassBuilder &PB, StringRef Params) {
-      Mode = BoundsCheckingPass::ReportingMode::MinRuntimeAbort;
+      Options.Mode = BoundsCheckingPass::ReportingMode::MinRuntimeAbort;
+    } else if (ParamName.consume_front("merge=")) {
+      if (ParamName == "true")
```
we don't do true/false here

usually if it's missing -> false, if it's present -> true

Check `ParamName == "kernel"` in this file

thurstond replied:
Changed to simple match of 'merge'

vitalybuka commented on llvm/lib/Transforms/Instrumentation/BoundsChecking.cpp:
```diff
@@ -299,7 +303,7 @@ void BoundsCheckingPass::printPipeline(
   static_cast<PassInfoMixin<BoundsCheckingPass> *>(this)->printPipeline(
       OS, MapClassName2PassName);
-  switch (Mode) {
+  switch (Options.Mode) {
```
you need to update this one function as well, this is opposite of parse

thurstond committed:
Address Vitaly's feedback

thurstond committed:
Update tests to add merge flag

vitalybuka commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -1208,6 +1202,11 @@ Sanitizers
 
+- ``-fsanitize-merge`` (default) and ``-fno-sanitize-merge`` have been added for
+  fine-grained, unified control of which UBSan checks can potentially be merged
+  by the compiler (for example,
```
Please mention on  https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#stack-traces-and-report-symbolization as well

thurstond committed:
Mention -fno-sanitize-merge in
https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#stack-traces-and-report-symbolization,
per Vitaly's feedback
